### ISSUE TITLE
Generate TypeScript types from types defined in schema.graphql

### ISSUE THREAD

**Issue by sapkra** [prisma/graphqlgen#181]
## Description
I got some problems with missing types & had to fixed them by defining them by myself.
I think an easier solution would be to genereate them automatically from the types defined in `schema.graphql`

## Additional context
More information in my old ticket #163.

**Comment by schickling** [comment]
Could you provide some more clarifying information please:

1) When you're saying "missing types": Are you referring to missing model definitions?
2) The current workflow for missing models is to print a suggestion in the CLI output. Are you aware of this behavior? Where would you suggest the missing models should be generated to?

**Comment by sapkra** [comment]
Yeah I mean exactly that what is generated in the CLI. And you were right, I was not aware of this behavior.

First of all `user` should not be an `any` I think.

schema.graphql
```graphql
type AuthPayload {
  token: String!
  user: User!
}
```
Generated model
```typescript
export interface AuthPayload {
  token: string
  user: any
}
```

My suggestion would be to just put them into the `generated/graphqlgen.ts` so that you also don't need to import them anymore.
But yeah it's definitely a design decision which depends on your vision of this project.

**Comment by kevinmarrec** [comment]
I agree with @sapkra that definining the `AuthPayload` in `types.ts` and refers to it in `graphqlgen.yml` to have finally `import { AuthPayload } from "../types";` in `generated/graphqlgen.ts` could be replaced by just having `graphqlgen` looking for "extra" models in `schema.graphql` which has not been referenced in `graphqlgen.yml` and do it by itself.

I mean actually if you forget to do it yourself, `graphqlgen` tells you "put this in your `types.ts`". So it's only about copying what `graphqlgen` already outputs inside `generated/graphqlgen.ts` instead of importing it from `../types`

**Comment by kevinmarrec** [comment]
@schickling Issue was referenced in #237 but I don't really understood why, and is there any thoughts about what me and @sapkra wrote ?

**Comment by kevinmarrec** [comment]
@schickling Any updates on this ?

Without automatically generating these models It's gonna be a pain for big projects to maintain both the Schema `schema.graphql` AND their respectives model definitions in a `types.ts`.

Exemple :

Auto generate 
```graphql
type Test {
  a: String
  b: String!
  c: Int
  d: Int!
}
```
into
```ts
export interface Test {
  a?: string
  b: string
  c?: number
  c: number
}
```

If it's not auto-generated, `graphqlgen` is requesting the user to copy/paste, why `graphqlgen` can't do it by itself ?

**Comment by emmenko** [comment]
I would also like to have the types being auto generated, I don't see a point in having to maintain those manually.

For example, I've been experimenting with using [`graphql-code-generator`](https://github.com/dotansimha/graphql-code-generator) to generate the types by doing an introspection query of the schema, then run `graphqlgen` to generate the types for resolvers etc.

**Comment by vjsingh** [comment]
I have literally the exact same use case, would be nice if this could be auto-generated

**Comment by kevinmarrec** [comment]
@schickling Any thoughts ?

**Comment by vjsingh** [comment]
Is the Payload pattern not used by the Prisma team? I'm curious about how structure input types for their mutations. The idea here is to create object for all inputs and outputs which makes them more maintainable in the future, since clients do not need to worry about getting all the arguments and their order, etc. correct

There is quite a lot of boilerplate currently for this pattern as seen below.

Is the recommendation simply not to use this pattern? Or is there another way of simplifying?

@nikolasburk @divyenduz 

schema.graphql: (ideally this would be all you need to write?)
```
addNote(input: AddNoteInput!): AddNotePayload!
input AddNoteInput {
  note: String!
  ...
}
type AddNotePayload {
  result: String
}
```

types.ts:
```
export interface AddNotePayload {
  result: string | null;
}
```

resolvers/AddNotePayload.ts:
```
import { AddNotePayloadResolvers } from '../generated/graphqlgen';

export const AddNotePayload: AddNotePayloadResolvers.Type = {
  ...AddNotePayloadResolvers.defaultResolvers,
};
```


resolvers/index.ts:
```
import { AddNotePayload } from './AddNotePayload';

export const resolvers: Resolvers = {
  ...
  AddNotePayload,
```

**Comment by koenpunt** [comment]
I think the use for auto generated models really depends on the project.
For example; I created a GraphQL gateway where the data being fetched from upstream servers, doesn't match with the GraphQL fields. So I would have no use for autogenerated models, because the models should be based on the shape of the upstream data. 
For the same reason I have no use for the generated `defaultResolvers`, hence why I opened an issue to be able to disable generation of them; #350.
That said, I'm not against autogenerating models, but it should be an option, not a default.

**Comment by jasonkuhrt** [comment]
Another use-case here https://github.com/prisma/graphqlgen/issues/436#issuecomment-462009269

**Comment by jasonkuhrt** [comment]
Need to decide:

1. which folder/file are model types generated into?

    Good start might be (roughly copying resolver types output file config):

    ```yml
    models:
      output: ./source/generated/models.ts
    ```

2. how are generated models merged with manually-created ones?

    - If a developer writes a same-named model, then don't generate one?
    - Or, if a developer writes a same-named model, then merge them via  [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html)? This would require users to write `interface Foo { ... }` instead of `type Foo { ... }`
    - allow developers to import the generated model types allowing them full control
    
2. how are relations generated?

    For example (https://github.com/prisma/graphqlgen/issues/448):

    ```ts
    type Post {
        id: ID!
        title: String!
        content: String!
        comments: [Comment!]
        user: User!
    }
    ```
    ```ts
    export interface Post {
        id: string;
        title: string;
        content: string;

        // ???
        comments: string[] // comment IDs that other resolvers will fetch data for
        user: string // user ID that another resolver will fetch data for
    }
    ```
